<a href="https://colab.research.google.com/github/bserikbek/ML_SmartAnalitica/blob/main/Beibit_Serikbek_Module_2_Homework_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Модуль 2. Домашнее задание 1**

1. Импортируйте библиотеку Pandas и покажите версию Pandas

In [2]:
import pandas as pd

pd.__version__

'2.2.2'

2. Импортируйте файл в датафрейм под названием "Electric_Vehicle_Population_Data.csv" из https://drive.google.com/file/d/172oRlGRIHck8CcuOHmLszw63VsfvigFg/view?usp=drive_link (вам необходимо скачать) и посмотрите информацию о датафрейме

In [3]:

df = pd.read_csv('Electric_Vehicle_Population_Data.csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 210165 entries, 0 to 210164
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         210165 non-null  object 
 1   County                                             210161 non-null  object 
 2   City                                               210161 non-null  object 
 3   State                                              210165 non-null  object 
 4   Postal Code                                        210161 non-null  float64
 5   Model Year                                         210165 non-null  int64  
 6   Make                                               210165 non-null  object 
 7   Model                                              210165 non-null  object 
 8   Electric Vehicle Type                              210165 non-null  object

3. Сколько строк и столбцов имеет датафрейм?

In [4]:
df.shape

(210165, 17)

Ответ: 210165 rows & 17 columns

4. Какие столбцы имеют пустые ячейки?

In [5]:
df.shape[1] - df[df.isnull().any(axis=1)].notnull().all(axis=0).sum()

9

Ответ: 9

5. Покажите верхние 6 строк датафрейма?

In [6]:
df.head(6)

VIN (1-10)     County       City State  Postal Code  Model Year     Make  \
0  5UXTA6C0XM     Kitsap    Seabeck    WA      98380.0        2021      BMW   
1  5YJ3E1EB1J     Kitsap    Poulsbo    WA      98370.0        2018    TESLA   
2  WP0AD2A73G  Snohomish    Bothell    WA      98012.0        2016  PORSCHE   
3  5YJ3E1EB5J     Kitsap  Bremerton    WA      98310.0        2018    TESLA   
4  1N4AZ1CP3K       King    Redmond    WA      98052.0        2019   NISSAN   
5  3FA6P0PU5F  Snohomish    Bothell    WA      98012.0        2015     FORD   

      Model                   Electric Vehicle Type  \
0        X5  Plug-in Hybrid Electric Vehicle (PHEV)   
1   MODEL 3          Battery Electric Vehicle (BEV)   
2  PANAMERA  Plug-in Hybrid Electric Vehicle (PHEV)   
3   MODEL 3          Battery Electric Vehicle (BEV)   
4      LEAF          Battery Electric Vehicle (BEV)   
5    FUSION  Plug-in Hybrid Electric Vehicle (PHEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible            30.0   
1           Clean Alternative Fuel Vehicle Eligible           215.0   
2             Not eligible due to low battery range            15.0   
3           Clean Alternative Fuel Vehicle Eligible           215.0   
4           Clean Alternative Fuel Vehicle Eligible           150.0   
5             Not eligible due to low battery range            19.0   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0        0.0                  35.0       267929112   
1        0.0                  23.0       475911439   
2        0.0                   1.0       101971278   
3        0.0                  23.0       474363746   
4        0.0                  45.0       476346482   
5        0.0                  21.0         9620047   

                  Vehicle Location  \
0  POINT (-122.8728334 47.5798304)   
1  POINT (-122.6368884 47.7469547)   
2    POINT (-122.206146 47.839957)   
3  POINT (-122.6231895 47.5930874)   
4      POINT (-122.13158 47.67858)   
5    POINT (-122.206146 47.839957)   

                                Electric Utility  2020 Census Tract  
0                         PUGET SOUND ENERGY INC       5.303509e+10  
1                         PUGET SOUND ENERGY INC       5.303509e+10  
2                         PUGET SOUND ENERGY INC       5.306105e+10  
3                         PUGET SOUND ENERGY INC       5.303508e+10  
4  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
5                         PUGET SOUND ENERGY INC       5.306104e+10

6. Покажите нижние 3 строк датафрейма?

In [7]:
df.tail(3)

VIN (1-10)   County              City State  Postal Code  Model Year  \
210162  JTMAB3FV3P   Pierce  University Place    WA      98466.0        2023   
210163  5YJ3E1EA1M     King          Kirkland    WA      98034.0        2021   
210164  7SAYGDEE0P  Spokane           Spokane    WA      99208.0        2023   

          Make       Model                   Electric Vehicle Type  \
210162  TOYOTA  RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
210163   TESLA     MODEL 3          Battery Electric Vehicle (BEV)   
210164   TESLA     MODEL Y          Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
210162            Clean Alternative Fuel Vehicle Eligible            42.0   
210163  Eligibility unknown as battery range has not b...             0.0   
210164  Eligibility unknown as battery range has not b...             0.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
210162        0.0                  28.0       254488797   
210163        0.0                   1.0       137279852   
210164        0.0                   6.0       241306740   

                       Vehicle Location  \
210162   POINT (-122.538546 47.2298959)   
210163  POINT (-122.2026532 47.7210518)   
210164  POINT (-117.4268937 47.7323627)   

                                         Electric Utility  2020 Census Tract  
210162  BONNEVILLE POWER ADMINISTRATION||CITY OF TACOM...       5.305307e+10  
210163      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
210164  BONNEVILLE POWER ADMINISTRATION||AVISTA CORP||...       5.306301e+10

7. Заполните пустые ячейки столбца Legislative District максимальным значением данного столбца

In [8]:
filterNaNLegis = df['Legislative District'].isnull()

display(df[filterNaNLegis])

df.fillna({'Legislative District': df['Legislative District'].max()}, inplace=True)

display(df[filterNaNLegis])


VIN (1-10)          County        City State  Postal Code  Model Year  \
40238   1GKB0RDCXR          Macomb      Warren    MI      48092.0        2024   
45209   1GKB0RDC0R          Macomb      Warren    MI      48092.0        2024   
53795   JTMAB3FV4N            Polk       Salem    OR      97304.0        2022   
54250   1GKB0RDC7R          Macomb      Warren    MI      48092.0        2024   
55076   3GNKDBRJXR          Macomb      Warren    MI      48092.0        2024   
...            ...             ...         ...   ...          ...         ...   
209613  2C4RC1S71M       San Diego     La Mesa    CA      91942.0        2021   
209898  JTMAB3FV2M        Muscogee    Columbus    GA      31907.0        2021   
209946  5YJYGDEE2M  San Bernardino  Fort Irwin    CA      92310.0        2021   
210118  7SAYGAEE5P       San Mateo    Atherton    CA      94027.0        2023   
210140  1G1RD6E4XE  Prince William    Manassas    VA      20111.0        2014   

             Make          Model                   Electric Vehicle Type  \
40238         GMC  HUMMER EV SUV          Battery Electric Vehicle (BEV)   
45209         GMC  HUMMER EV SUV          Battery Electric Vehicle (BEV)   
53795      TOYOTA     RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
54250         GMC  HUMMER EV SUV          Battery Electric Vehicle (BEV)   
55076   CHEVROLET      BLAZER EV          Battery Electric Vehicle (BEV)   
...           ...            ...                                     ...   
209613   CHRYSLER       PACIFICA  Plug-in Hybrid Electric Vehicle (PHEV)   
209898     TOYOTA     RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
209946      TESLA        MODEL Y          Battery Electric Vehicle (BEV)   
210118      TESLA        MODEL Y          Battery Electric Vehicle (BEV)   
210140  CHEVROLET           VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
40238   Eligibility unknown as battery range has not b...             0.0   
45209   Eligibility unknown as battery range has not b...             0.0   
53795             Clean Alternative Fuel Vehicle Eligible            42.0   
54250   Eligibility unknown as battery range has not b...             0.0   
55076   Eligibility unknown as battery range has not b...             0.0   
...                                                   ...             ...   
209613            Clean Alternative Fuel Vehicle Eligible            32.0   
209898            Clean Alternative Fuel Vehicle Eligible            42.0   
209946  Eligibility unknown as battery range has not b...             0.0   
210118  Eligibility unknown as battery range has not b...             0.0   
210140            Clean Alternative Fuel Vehicle Eligible            38.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
40238         0.0                   NaN       273369309   
45209         0.0                   NaN       269827470   
53795         0.0                   NaN       218086949   
54250         0.0                   NaN       266636695   
55076         0.0                   NaN       269438851   
...           ...                   ...             ...   
209613        0.0                   NaN       262847308   
209898        0.0                   NaN       208633246   
209946        0.0                   NaN       142870444   
210118        0.0                   NaN       228732409   
210140        0.0                   NaN       217175514   

                       Vehicle Location  \
40238      POINT (-83.064593 42.512487)   
45209      POINT (-83.064593 42.512487)   
53795   POINT (-123.0771613 44.9631335)   
54250      POINT (-83.064593 42.512487)   
55076      POINT (-83.064593 42.512487)   
...                                 ...   
209613  POINT (-117.0230197 32.7779798)   
209898     POINT (-84.902073 32.485872)   
209946  POINT (-116.6984466 35.2618843)   
210118  POINT (-122.1941982 37.4534418)   
210140       PO

VIN (1-10)          County        City State  Postal Code  Model Year  \
40238   1GKB0RDCXR          Macomb      Warren    MI      48092.0        2024   
45209   1GKB0RDC0R          Macomb      Warren    MI      48092.0        2024   
53795   JTMAB3FV4N            Polk       Salem    OR      97304.0        2022   
54250   1GKB0RDC7R          Macomb      Warren    MI      48092.0        2024   
55076   3GNKDBRJXR          Macomb      Warren    MI      48092.0        2024   
...            ...             ...         ...   ...          ...         ...   
209613  2C4RC1S71M       San Diego     La Mesa    CA      91942.0        2021   
209898  JTMAB3FV2M        Muscogee    Columbus    GA      31907.0        2021   
209946  5YJYGDEE2M  San Bernardino  Fort Irwin    CA      92310.0        2021   
210118  7SAYGAEE5P       San Mateo    Atherton    CA      94027.0        2023   
210140  1G1RD6E4XE  Prince William    Manassas    VA      20111.0        2014   

             Make          Model                   Electric Vehicle Type  \
40238         GMC  HUMMER EV SUV          Battery Electric Vehicle (BEV)   
45209         GMC  HUMMER EV SUV          Battery Electric Vehicle (BEV)   
53795      TOYOTA     RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
54250         GMC  HUMMER EV SUV          Battery Electric Vehicle (BEV)   
55076   CHEVROLET      BLAZER EV          Battery Electric Vehicle (BEV)   
...           ...            ...                                     ...   
209613   CHRYSLER       PACIFICA  Plug-in Hybrid Electric Vehicle (PHEV)   
209898     TOYOTA     RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
209946      TESLA        MODEL Y          Battery Electric Vehicle (BEV)   
210118      TESLA        MODEL Y          Battery Electric Vehicle (BEV)   
210140  CHEVROLET           VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
40238   Eligibility unknown as battery range has not b...             0.0   
45209   Eligibility unknown as battery range has not b...             0.0   
53795             Clean Alternative Fuel Vehicle Eligible            42.0   
54250   Eligibility unknown as battery range has not b...             0.0   
55076   Eligibility unknown as battery range has not b...             0.0   
...                                                   ...             ...   
209613            Clean Alternative Fuel Vehicle Eligible            32.0   
209898            Clean Alternative Fuel Vehicle Eligible            42.0   
209946  Eligibility unknown as battery range has not b...             0.0   
210118  Eligibility unknown as battery range has not b...             0.0   
210140            Clean Alternative Fuel Vehicle Eligible            38.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
40238         0.0                  49.0       273369309   
45209         0.0                  49.0       269827470   
53795         0.0                  49.0       218086949   
54250         0.0                  49.0       266636695   
55076         0.0                  49.0       269438851   
...           ...                   ...             ...   
209613        0.0                  49.0       262847308   
209898        0.0                  49.0       208633246   
209946        0.0                  49.0       142870444   
210118        0.0                  49.0       228732409   
210140        0.0                  49.0       217175514   

                       Vehicle Location  \
40238      POINT (-83.064593 42.512487)   
45209      POINT (-83.064593 42.512487)   
53795   POINT (-123.0771613 44.9631335)   
54250      POINT (-83.064593 42.512487)   
55076      POINT (-83.064593 42.512487)   
...                                 ...   
209613  POINT (-117.0230197 32.7779798)   
209898     POINT (-84.902073 32.485872)   
209946  POINT (-116.6984466 35.2618843)   
210118  POINT (-122.1941982 37.4534418)   
210140       PO

8. Заполните пустые ячейки столбца County значением 'No data'

In [9]:
filterNaNCounty = df['County'].isnull()

display(df[filterNaNCounty])

df.fillna({'County': 'No data'}, inplace= True)

display(df[filterNaNCounty])

VIN (1-10) County City State  Postal Code  Model Year       Make  \
117629  5YJ3E1EB2M    NaN  NaN    NS          NaN        2021      TESLA   
148153  5YJXCAE24H    NaN  NaN    BC          NaN        2017      TESLA   
180881  WBAJA9C50K    NaN  NaN    AE          NaN        2019        BMW   
205349  1G1RB6S53J    NaN  NaN    BC          NaN        2018  CHEVROLET   

          Model                   Electric Vehicle Type  \
117629  MODEL 3          Battery Electric Vehicle (BEV)   
148153  MODEL X          Battery Electric Vehicle (BEV)   
180881     530E  Plug-in Hybrid Electric Vehicle (PHEV)   
205349     VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
117629  Eligibility unknown as battery range has not b...             0.0   
148153            Clean Alternative Fuel Vehicle Eligible           200.0   
180881              Not eligible due to low battery range            16.0   
205349            Clean Alternative Fuel Vehicle Eligible            53.0   

        Base MSRP  Legislative District  DOL Vehicle ID Vehicle Location  \
117629        0.0                  49.0       179569743              NaN   
148153        0.0                  49.0       159850029              NaN   
180881    53400.0                  49.0       244582593              NaN   
205349        0.0                  49.0       477613216              NaN   

       Electric Utility  2020 Census Tract  
117629              NaN                NaN  
148153              NaN                NaN  
180881              NaN                NaN  
205349              NaN                NaN

VIN (1-10)   County City State  Postal Code  Model Year       Make  \
117629  5YJ3E1EB2M  No data  NaN    NS          NaN        2021      TESLA   
148153  5YJXCAE24H  No data  NaN    BC          NaN        2017      TESLA   
180881  WBAJA9C50K  No data  NaN    AE          NaN        2019        BMW   
205349  1G1RB6S53J  No data  NaN    BC          NaN        2018  CHEVROLET   

          Model                   Electric Vehicle Type  \
117629  MODEL 3          Battery Electric Vehicle (BEV)   
148153  MODEL X          Battery Electric Vehicle (BEV)   
180881     530E  Plug-in Hybrid Electric Vehicle (PHEV)   
205349     VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
117629  Eligibility unknown as battery range has not b...             0.0   
148153            Clean Alternative Fuel Vehicle Eligible           200.0   
180881              Not eligible due to low battery range            16.0   
205349            Clean Alternative Fuel Vehicle Eligible            53.0   

        Base MSRP  Legislative District  DOL Vehicle ID Vehicle Location  \
117629        0.0                  49.0       179569743              NaN   
148153        0.0                  49.0       159850029              NaN   
180881    53400.0                  49.0       244582593              NaN   
205349        0.0                  49.0       477613216              NaN   

       Electric Utility  2020 Census Tract  
117629              NaN                NaN  
148153              NaN                NaN  
180881              NaN                NaN  
205349              NaN                NaN

In [10]:
df[~df['Base MSRP'].isin([0])].sort_values('Base MSRP')

VIN (1-10)     County           City State  Postal Code  Model Year  \
107329  KNDJP3AE2G       King        Redmond    WA      98053.0        2016   
154861  KNDJX3AE0G       King       Issaquah    WA      98027.0        2016   
59950   KNDJX3AE3G     Pierce         Tacoma    WA      98406.0        2016   
154043  KNDJX3AE4G       King        Seattle    WA      98119.0        2016   
153944  KNDJX3AEXG  Snohomish        Edmonds    WA      98026.0        2016   
...            ...        ...            ...   ...          ...         ...   
7440    WP1BM2AY4R       King       Kirkland    WA      98033.0        2024   
23186   WP1BM2AY1R       King        Seattle    WA      98108.0        2024   
64653   WP1BM2AY1R       King  Mercer Island    WA      98040.0        2024   
122662  WP1BM2AY8R    Spokane   Liberty Lake    WA      99019.0        2024   
141220  WP1BM2AYXR       King         Renton    WA      98058.0        2024   

           Make    Model                   Electric Vehicle Type  \
107329      KIA     SOUL          Battery Electric Vehicle (BEV)   
154861      KIA     SOUL          Battery Electric Vehicle (BEV)   
59950       KIA     SOUL          Battery Electric Vehicle (BEV)   
154043      KIA     SOUL          Battery Electric Vehicle (BEV)   
153944      KIA     SOUL          Battery Electric Vehicle (BEV)   
...         ...      ...                                     ...   
7440    PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
23186   PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
64653   PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
122662  PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
141220  PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   

       Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
107329           Clean Alternative Fuel Vehicle Eligible            93.0   
154861           Clean Alternative Fuel Vehicle Eligible            93.0   
59950            Clean Alternative Fuel Vehicle Eligible            93.0   
154043           Clean Alternative Fuel Vehicle Eligible            93.0   
153944           Clean Alternative Fuel Vehicle Eligible            93.0   
...                                                  ...             ...   
7440             Clean Alternative Fuel Vehicle Eligible             NaN   
23186            Clean Alternative Fuel Vehicle Eligible             NaN   
64653            Clean Alternative Fuel Vehicle Eligible             NaN   
122662           Clean Alternative Fuel Vehicle Eligible             NaN   
141220           Clean Alternative Fuel Vehicle Eligible             NaN   

        Base MSRP  Legislative District  DOL Vehicle ID  \
107329    31950.0                  45.0         3654696   
154861    31950.0                   5.0       208154754   
59950     31950.0                  27.0       140345614   
154043    31950.0                  36.0       345277586   
153944    31950.0                  32.0       227676676   
...           ...                   ...             ...   
7440          NaN                  45.0       270889363   
23186         NaN                  11.0       272220189   
64653         NaN                  41.0       272497872   
122662        NaN                   4.0       270064791   
141220        NaN                  11.0       271748004   

                       Vehicle Location  \
107329    POINT (-122.030684 47.669552)   
154861  POINT (-122.0402383 47.5276876)   
59950     POINT (-122.498503 47.262687)   
154043       POINT (-122.3684 47.64586)   
153944    POINT (-122.338942 47.821454)   
...                                 ...   
7440     POINT (-122.1925969 47.676241)   
23186   POINT (-122.3173531 47.5484673)   
64653   POINT (-122.2283088 47.5737238)   
122662    POINT (-117.09883 47.6679943)   
141220  POINT (-122.1476337 47.4438471)   

                                         Electric Utility  2020 Census Tract  
107329      PUGET SOUND 

9. Заполните пустые ячейки столбца  Electric Range средним значением  столбца Base MSRP

In [11]:
filterNaNElectric = df['Electric Range'].isnull()

display(df[filterNaNElectric])

df.fillna({'Electric Range': df['Base MSRP'].mean()}, inplace= True)

display(df[filterNaNElectric])

VIN (1-10)   County           City State  Postal Code  Model Year  \
7440    WP1BM2AY4R     King       Kirkland    WA      98033.0        2024   
23186   WP1BM2AY1R     King        Seattle    WA      98108.0        2024   
64653   WP1BM2AY1R     King  Mercer Island    WA      98040.0        2024   
122662  WP1BM2AY8R  Spokane   Liberty Lake    WA      99019.0        2024   
141220  WP1BM2AYXR     King         Renton    WA      98058.0        2024   

           Make    Model                   Electric Vehicle Type  \
7440    PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
23186   PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
64653   PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
122662  PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
141220  PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   

       Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
7440             Clean Alternative Fuel Vehicle Eligible             NaN   
23186            Clean Alternative Fuel Vehicle Eligible             NaN   
64653            Clean Alternative Fuel Vehicle Eligible             NaN   
122662           Clean Alternative Fuel Vehicle Eligible             NaN   
141220           Clean Alternative Fuel Vehicle Eligible             NaN   

        Base MSRP  Legislative District  DOL Vehicle ID  \
7440          NaN                  45.0       270889363   
23186         NaN                  11.0       272220189   
64653         NaN                  41.0       272497872   
122662        NaN                   4.0       270064791   
141220        NaN                  11.0       271748004   

                       Vehicle Location  \
7440     POINT (-122.1925969 47.676241)   
23186   POINT (-122.3173531 47.5484673)   
64653   POINT (-122.2283088 47.5737238)   
122662    POINT (-117.09883 47.6679943)   
141220  POINT (-122.1476337 47.4438471)   

                                         Electric Utility  2020 Census Tract  
7440        PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
23186        CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
64653       PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
122662  BONNEVILLE POWER ADMINISTRATION||AVISTA CORP||...       5.306301e+10  
141220      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10

VIN (1-10)   County           City State  Postal Code  Model Year  \
7440    WP1BM2AY4R     King       Kirkland    WA      98033.0        2024   
23186   WP1BM2AY1R     King        Seattle    WA      98108.0        2024   
64653   WP1BM2AY1R     King  Mercer Island    WA      98040.0        2024   
122662  WP1BM2AY8R  Spokane   Liberty Lake    WA      99019.0        2024   
141220  WP1BM2AYXR     King         Renton    WA      98058.0        2024   

           Make    Model                   Electric Vehicle Type  \
7440    PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
23186   PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
64653   PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
122662  PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
141220  PORSCHE  CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   

       Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
7440             Clean Alternative Fuel Vehicle Eligible      897.676889   
23186            Clean Alternative Fuel Vehicle Eligible      897.676889   
64653            Clean Alternative Fuel Vehicle Eligible      897.676889   
122662           Clean Alternative Fuel Vehicle Eligible      897.676889   
141220           Clean Alternative Fuel Vehicle Eligible      897.676889   

        Base MSRP  Legislative District  DOL Vehicle ID  \
7440          NaN                  45.0       270889363   
23186         NaN                  11.0       272220189   
64653         NaN                  41.0       272497872   
122662        NaN                   4.0       270064791   
141220        NaN                  11.0       271748004   

                       Vehicle Location  \
7440     POINT (-122.1925969 47.676241)   
23186   POINT (-122.3173531 47.5484673)   
64653   POINT (-122.2283088 47.5737238)   
122662    POINT (-117.09883 47.6679943)   
141220  POINT (-122.1476337 47.4438471)   

                                         Electric Utility  2020 Census Tract  
7440        PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
23186        CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
64653       PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
122662  BONNEVILLE POWER ADMINISTRATION||AVISTA CORP||...       5.306301e+10  
141220      PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10

10. Теперь после заполнения пустых ячеек по 2-м вышеуказанным столбцам, неоходимо удалить все строки, которые содержат пустые ячейка

In [16]:
display(df[df.isnull().any(axis=1)])

df.dropna(inplace= True)

print(df[df.isnull().any(axis=1)])

VIN (1-10)   County           City State  Postal Code  Model Year  \
7440    WP1BM2AY4R     King       Kirkland    WA      98033.0        2024   
23186   WP1BM2AY1R     King        Seattle    WA      98108.0        2024   
64653   WP1BM2AY1R     King  Mercer Island    WA      98040.0        2024   
117629  5YJ3E1EB2M  No data            NaN    NS          NaN        2021   
122662  WP1BM2AY8R  Spokane   Liberty Lake    WA      99019.0        2024   
141220  WP1BM2AYXR     King         Renton    WA      98058.0        2024   
148153  5YJXCAE24H  No data            NaN    BC          NaN        2017   
156934  1C4JJXP69P  Pacific     Long Beach    WA      98634.0        2023   
162925  1G1FW6S08N  Pacific     Long Beach    WA      98634.0        2022   
174901  KMHE54L2XH  Pacific     Long Beach    WA      98634.0        2017   
180881  WBAJA9C50K  No data            NaN    AE          NaN        2019   
190936  JTDKARFP1H  Pacific     Long Beach    WA      98634.0        2017   
198123  7SAYGDEE0N  Pacific     Long Beach    WA      98634.0        2022   
201296  KNDJX3AE8H  Pacific     Long Beach    WA      98634.0        2017   
205349  1G1RB6S53J  No data            NaN    BC          NaN        2018   

             Make        Model                   Electric Vehicle Type  \
7440      PORSCHE      CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
23186     PORSCHE      CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
64653     PORSCHE      CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
117629      TESLA      MODEL 3          Battery Electric Vehicle (BEV)   
122662    PORSCHE      CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
141220    PORSCHE      CAYENNE  Plug-in Hybrid Electric Vehicle (PHEV)   
148153      TESLA      MODEL X          Battery Electric Vehicle (BEV)   
156934       JEEP     WRANGLER  Plug-in Hybrid Electric Vehicle (PHEV)   
162925  CHEVROLET      BOLT EV          Battery Electric Vehicle (BEV)   
174901    HYUNDAI       SONATA  Plug-in Hybrid Electric Vehicle (PHEV)   
180881        BMW         530E  Plug-in Hybrid Electric Vehicle (PHEV)   
190936     TOYOTA  PRIUS PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
198123      TESLA      MODEL Y          Battery Electric Vehicle (BEV)   
201296        KIA      SOUL EV          Battery Electric Vehicle (BEV)   
205349  CHEVROLET         VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
7440              Clean Alternative Fuel Vehicle Eligible      897.676889   
23186             Clean Alternative Fuel Vehicle Eligible      897.676889   
64653             Clean Alternative Fuel Vehicle Eligible      897.676889   
117629  Eligibility unknown as battery range has not b...        0.000000   
122662            Clean Alternative Fuel Vehicle Eligible      897.676889   
141220            Clean Alternative Fuel Vehicle Eligible      897.676889   
148153            Clean Alternative Fuel Vehicle Eligible      200.000000   
156934              Not eligible due to low battery range       21.000000   
162925  Eligibility unknown as battery range has not b...        0.000000   
174901              Not eligible due to low battery range       27.000000   
180881              Not eligible due to low battery range       16.000000   
190936              Not eligible due to low battery range       25.000000   
198123  Eligibility unknown as battery range has not b...        0.000000   
201296            Clean Alternative Fuel Vehicle Eligible       93.000000   
205349            Clean Alternative Fuel Vehicle Eligible       53.000000   

        Base MSRP  Legislative District  DOL Vehicle ID  \
7440          NaN                  45.0       270889363   
23186         NaN                  11.0       272220189   
64653         NaN                  41.0       272497872   
117629        0.0                  49.0       179569743   
122662        NaN                   4.0       270064791   
141

Empty DataFrame
Columns: [VIN (1-10), County, City, State, Postal Code, Model Year, Make, Model, Electric Vehicle Type, Clean Alternative Fuel Vehicle (CAFV) Eligibility, Electric Range, Base MSRP, Legislative District, DOL Vehicle ID, Vehicle Location, Electric Utility, 2020 Census Tract]
Index: []


11. Используя индексы датафрейма с 20 по 150, вытащите данные по столбцам 'VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year',
       'Make', 'Model', 'Electric Vehicle Type'. Результат запишите на новый датафрейм

In [20]:
df.shape

df_20_150 = df[['VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year', 'Make', 'Model', 'Electric Vehicle Type']].loc[20:150]


12. Покажите кол-во строк и столбцов используя .shape

In [21]:
df_20_150.shape

(131, 9)

13. Используя функцию Pandas покажите статистиечкие результаты основного датафрейма

In [23]:
df.describe()

Postal Code     Model Year  Electric Range      Base MSRP  \
count  210150.000000  210150.000000   210150.000000  210150.000000   
mean    98178.194647    2021.048670       50.602579     897.312039   
std      2445.491444       2.988946       86.974319    7652.606144   
min      1731.000000    1999.000000        0.000000       0.000000   
25%     98052.000000    2019.000000        0.000000       0.000000   
50%     98125.000000    2022.000000        0.000000       0.000000   
75%     98373.000000    2023.000000       42.000000       0.000000   
max     99577.000000    2025.000000      337.000000  845000.000000   

       Legislative District  DOL Vehicle ID  2020 Census Tract  
count         210150.000000    2.101500e+05       2.101500e+05  
mean              28.972510    2.290765e+08       5.297929e+10  
std               14.921063    7.115445e+07       1.551507e+09  
min                1.000000    4.469000e+03       1.001020e+09  
25%               17.000000    1.948825e+08       5.303301e+10  
50%               33.000000    2.405161e+08       5.303303e+10  
75%               42.000000    2.629754e+08       5.305307e+10  
max               49.000000    4.792548e+08       5.602100e+10

14. Из полученного стат результата, какой столбец имеет 75% значение 0?

In [66]:
import numpy as np

numeric_df = df.select_dtypes(include=[np.number])

numeric_df.apply(lambda col: np.percentile(col, 75))


Postal Code             9.837300e+04
Model Year              2.023000e+03
Electric Range          4.200000e+01
Base MSRP               0.000000e+00
Legislative District    4.200000e+01
DOL Vehicle ID          2.629754e+08
2020 Census Tract       5.305307e+10
dtype: float64

Ответ: Base MSRP

15. Покажите корреляцию между столбцами 'Base MSRP', 'Electric Range'

In [65]:

df[['Base MSRP', 'Electric Range']].corr()

Base MSRP  Electric Range
Base MSRP        1.000000        0.114177
Electric Range   0.114177        1.000000

16. Что думаете о результате корреляции?

Ответ: Correlation (positive or negative) between those two columns is pretty weak that gives us insight that they don't have relation between each other. In other words, changes in one of them don't reflect to second column in the same or opposite way (does not have any impact at all).

17.  Необходимо отсортировать от меньшего к большеми по столбцу Model Year и перезапишите датафреим

In [69]:
df.sort_values(by = 'Model Year', inplace= True)

display(df)

df_20_150.sort_values(by = 'Model Year', inplace = True)

display(df_20_150)

VIN (1-10)     County           City State  Postal Code  Model Year  \
155362  1FTZR0813X     Pierce         Graham    WA      98338.0        1999   
118405  1FTZR0818X   San Juan      Eastsound    WA      98245.0        1999   
159077  1FTZR0813Y    Whatcom        Everson    WA      98247.0        2000   
135280  1FTZR0870Y   San Juan  Friday Harbor    WA      98250.0        2000   
42995   1FTZR0817Y       King        Seattle    WA      98117.0        2000   
...            ...        ...            ...   ...          ...         ...   
140517  7PDSGBBA2S       King       Kirkland    WA      98033.0        2025   
53176   JTJKKCFZ1S       King        Seattle    WA      98115.0        2025   
195152  5UX43EU08S  Snohomish     Mill Creek    WA      98012.0        2025   
159003  7PDSGBBA4S       King       Kirkland    WA      98034.0        2025   
200860  KNDRMDJH2S       King           Kent    WA      98042.0        2025   

          Make    Model                   Electric Vehicle Type  \
155362    FORD   RANGER          Battery Electric Vehicle (BEV)   
118405    FORD   RANGER          Battery Electric Vehicle (BEV)   
159077    FORD   RANGER          Battery Electric Vehicle (BEV)   
135280    FORD   RANGER          Battery Electric Vehicle (BEV)   
42995     FORD   RANGER          Battery Electric Vehicle (BEV)   
...        ...      ...                                     ...   
140517  RIVIAN      R1S          Battery Electric Vehicle (BEV)   
53176    LEXUS       NX  Plug-in Hybrid Electric Vehicle (PHEV)   
195152     BMW       X5  Plug-in Hybrid Electric Vehicle (PHEV)   
159003  RIVIAN      R1S          Battery Electric Vehicle (BEV)   
200860     KIA  SORENTO  Plug-in Hybrid Electric Vehicle (PHEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
155362            Clean Alternative Fuel Vehicle Eligible            74.0   
118405            Clean Alternative Fuel Vehicle Eligible            74.0   
159077            Clean Alternative Fuel Vehicle Eligible            58.0   
135280            Clean Alternative Fuel Vehicle Eligible            58.0   
42995             Clean Alternative Fuel Vehicle Eligible            58.0   
...                                                   ...             ...   
140517  Eligibility unknown as battery range has not b...             0.0   
53176             Clean Alternative Fuel Vehicle Eligible            37.0   
195152            Clean Alternative Fuel Vehicle Eligible            40.0   
159003  Eligibility unknown as battery range has not b...             0.0   
200860            Clean Alternative Fuel Vehicle Eligible            30.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
155362        0.0                   2.0       169463255   
118405        0.0                  40.0       262410760   
159077        0.0                  42.0         8650316   
135280        0.0                  40.0       221693048   
42995         0.0                  36.0       142611402   
...           ...                   ...             ...   
140517        0.0                  45.0       270348396   
53176         0.0                  46.0       270150219   
195152        0.0                  44.0       267106115   
159003        0.0                   1.0       273104485   
200860        0.0                   5.0       273289845   

                       Vehicle Location  \
155362  POINT (-122.2894997 47.0293015)   
118405  POINT (-122.9129975 48.6692106)   
159077  POINT (-122.3616332 48.9071317)   
135280  POINT (-123.0679454 48.5327669)   
42995     POINT (-122.377938 47.690166)   
...                                 ...   
140517   POINT (-122.1925969 47.676241)   
53176   POINT (-122.3008235 47.6862671)   
195152    POINT (-122.206146 47.839957)   
159003  POINT (-122.2026532 47.7210518)   
200860  POINT (-122.1145138 47.3581107)   

                                         Electric Utility  2020 Census Tract  
155362  BONNEVILLE POWER

VIN (1-10)     County      City State  Postal Code  Model Year  \
93   JN1AZ0CPXB  Snohomish   Everett    WA      98204.0        2011   
110  JN1AZ0CP7B   Thurston     Lacey    WA      98503.0        2011   
141  1G1RD6E43C     Yakima    Yakima    WA      98908.0        2012   
131  JN1AZ0CP6C     Island  Freeland    WA      98249.0        2012   
118  JN1AZ0CP5C    Stevens  Colville    WA      99114.0        2012   
..          ...        ...       ...   ...          ...         ...   
130  KMUKEDSBXR  Snohomish   Edmonds    WA      98020.0        2024   
149  JTMCB3FV0R    Stevens  Colville    WA      99114.0        2024   
43   5UX43EU04S     Yakima    Yakima    WA      98902.0        2025   
129  5UX43EU04S     Kitsap   Seabeck    WA      98380.0        2025   
82   1N4CZ1CV3S     Yakima    Zillah    WA      98953.0        2025   

          Make       Model                   Electric Vehicle Type  
93      NISSAN        LEAF          Battery Electric Vehicle (BEV)  
110     NISSAN        LEAF          Battery Electric Vehicle (BEV)  
141  CHEVROLET        VOLT  Plug-in Hybrid Electric Vehicle (PHEV)  
131     NISSAN        LEAF          Battery Electric Vehicle (BEV)  
118     NISSAN        LEAF          Battery Electric Vehicle (BEV)  
..         ...         ...                                     ...  
130    GENESIS        GV60          Battery Electric Vehicle (BEV)  
149     TOYOTA  RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)  
43         BMW          X5  Plug-in Hybrid Electric Vehicle (PHEV)  
129        BMW          X5  Plug-in Hybrid Electric Vehicle (PHEV)  
82      NISSAN        LEAF          Battery Electric Vehicle (BEV)  

[131 rows x 9 columns]

18. Создайте функцию, где смотря столбец 'Electric Range', необходимо заполнить новый столбец 'Electric Range Mapped' по условиям, если 'Electric Range' меньше 100 то 'Electric Range Mapped' равно 'Low', от 100 до 250 'Medium', выше 250 'High'

In [88]:
def distanceRange(erange):
  if erange < 100:
    return 'Low'
  elif erange <= 250:
    return 'Medium'
  else: return 'High'


df['Electric Range Mapped'] = df['Electric Range'].map(distanceRange)

df

Medium


VIN (1-10)     County           City State  Postal Code  Model Year  \
155362  1FTZR0813X     Pierce         Graham    WA      98338.0        1999   
118405  1FTZR0818X   San Juan      Eastsound    WA      98245.0        1999   
159077  1FTZR0813Y    Whatcom        Everson    WA      98247.0        2000   
135280  1FTZR0870Y   San Juan  Friday Harbor    WA      98250.0        2000   
42995   1FTZR0817Y       King        Seattle    WA      98117.0        2000   
...            ...        ...            ...   ...          ...         ...   
140517  7PDSGBBA2S       King       Kirkland    WA      98033.0        2025   
53176   JTJKKCFZ1S       King        Seattle    WA      98115.0        2025   
195152  5UX43EU08S  Snohomish     Mill Creek    WA      98012.0        2025   
159003  7PDSGBBA4S       King       Kirkland    WA      98034.0        2025   
200860  KNDRMDJH2S       King           Kent    WA      98042.0        2025   

          Make    Model                   Electric Vehicle Type  \
155362    FORD   RANGER          Battery Electric Vehicle (BEV)   
118405    FORD   RANGER          Battery Electric Vehicle (BEV)   
159077    FORD   RANGER          Battery Electric Vehicle (BEV)   
135280    FORD   RANGER          Battery Electric Vehicle (BEV)   
42995     FORD   RANGER          Battery Electric Vehicle (BEV)   
...        ...      ...                                     ...   
140517  RIVIAN      R1S          Battery Electric Vehicle (BEV)   
53176    LEXUS       NX  Plug-in Hybrid Electric Vehicle (PHEV)   
195152     BMW       X5  Plug-in Hybrid Electric Vehicle (PHEV)   
159003  RIVIAN      R1S          Battery Electric Vehicle (BEV)   
200860     KIA  SORENTO  Plug-in Hybrid Electric Vehicle (PHEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
155362            Clean Alternative Fuel Vehicle Eligible            74.0   
118405            Clean Alternative Fuel Vehicle Eligible            74.0   
159077            Clean Alternative Fuel Vehicle Eligible            58.0   
135280            Clean Alternative Fuel Vehicle Eligible            58.0   
42995             Clean Alternative Fuel Vehicle Eligible            58.0   
...                                                   ...             ...   
140517  Eligibility unknown as battery range has not b...             0.0   
53176             Clean Alternative Fuel Vehicle Eligible            37.0   
195152            Clean Alternative Fuel Vehicle Eligible            40.0   
159003  Eligibility unknown as battery range has not b...             0.0   
200860            Clean Alternative Fuel Vehicle Eligible            30.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
155362        0.0                   2.0       169463255   
118405        0.0                  40.0       262410760   
159077        0.0                  42.0         8650316   
135280        0.0                  40.0       221693048   
42995         0.0                  36.0       142611402   
...           ...                   ...             ...   
140517        0.0                  45.0       270348396   
53176         0.0                  46.0       270150219   
195152        0.0                  44.0       267106115   
159003        0.0                   1.0       273104485   
200860        0.0                   5.0       273289845   

                       Vehicle Location  \
155362  POINT (-122.2894997 47.0293015)   
118405  POINT (-122.9129975 48.6692106)   
159077  POINT (-122.3616332 48.9071317)   
135280  POINT (-123.0679454 48.5327669)   
42995     POINT (-122.377938 47.690166)   
...                                 ...   
140517   POINT (-122.1925969 47.676241)   
53176   POINT (-122.3008235 47.6862671)   
195152    POINT (-122.206146 47.839957)   
159003  POINT (-122.2026532 47.7210518)   
200860  POINT (-122.1145138 47.3581107)   

                                         Electric Utility  2020 Census Tract  \
155362  BONNEVILLE POWE

19. Покажите все строки индекса 1532 с помощью метода .loc

In [89]:
df.loc[1532]

VIN (1-10)                                                                                  KM8KNDA80S
County                                                                                            King
City                                                                                            Burien
State                                                                                               WA
Postal Code                                                                                    98168.0
Model Year                                                                                        2025
Make                                                                                           HYUNDAI
Model                                                                                        IONIQ 5 N
Electric Vehicle Type                                                   Battery Electric Vehicle (BEV)
Clean Alternative Fuel Vehicle (CAFV) Eligibility    Eligibility unknown as battery range has not b...
Electric Range                                                                                     0.0
Base MSRP                                                                                          0.0
Legislative District                                                                              33.0
DOL Vehicle ID                                                                               270066149
Vehicle Location                                                         POINT (-122.297205 47.486649)
Electric Utility                                          CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
2020 Census Tract                                                                        53033027000.0
Electric Range Mapped                                                                              Low
Name: 1532, dtype: object

20. Сохраните датафреим на формат csv и xlsx(Excel). Наименование файла можете дать по желанию

In [90]:
df.to_csv('output_bserikbek_df.csv')

In [92]:
df.to_excel('output_bserikbek_df.xlsx')